django/test/testcases.py

```diff
@@ -150,6 +150,48 @@ def __call__(self):
         )
 
 
+class _AssertSignalSentContext(object):
+    def __init__(self, test_case, signal, required_kwargs=None):
+        self.test_case = test_case
+        self.signal = signal
+        self.required_kwargs = required_kwargs
+
+    def _listener(self, sender, **kwargs):
+        self.signal_sent = True
+        self.received_kwargs = kwargs
+        self.sender = sender
+
+    def __enter__(self):
+        self.signal_sent = False
+        self.received_kwargs = {}
+        self.sender = None
+        self.signal.connect(self._listener)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.signal.disconnect(self._listener)
+        if not self.signal_sent:
+            self.test_case.fail('Signal was not sent.')
```

Comment:
@timgraham that's neat but that looks really fragile. Think `return Signal(providing_args=["app_config", "verbosity", "interactive", "using", "apps", "plan"])` where the name would end up being `'return Signal(providing_args'`) which can lead to more confusion than the actual situation.

We could also make `assertSignalSent` accept a `msg` argument (like other `assert` methods do) to allow the user to disambiguate the origin of the failure.

From my point of view the traceback is explicit enough to point the users at the correct location in their code base and figure out which signal was sent or not.